File: src/main/java/org/cerion/stocklist/overlays/IchimokuClouds.java
package org.cerion.stocklist.overlays;

import org.cerion.stocklist.PriceList;
import org.cerion.stocklist.arrays.FloatArray;
import org.cerion.stocklist.arrays.PairArray;
import org.cerion.stocklist.functions.types.PriceOverlay;

public class IchimokuClouds extends PriceOverlayBase {

    public IchimokuClouds() {
        super(PriceOverlay.CLOUD, 9, 26, 52);
    }

    public IchimokuClouds(Number ...params) {
        this();
        setParams(params);
    }

    @Override
    public PairArray eval(PriceList list) {
        return ichimokuCloud(list, getInt(0), getInt(1), getInt(2));
    }

    @Override
    public String getName() {
        return "Ichimoku Clouds";
    }

    private static PairArray ichimokuCloud(PriceList list, int conversion, int base, int span)
    {
        int size = list.size();
        FloatArray spanA = new FloatArray(size);
        FloatArray spanB = new FloatArray(size);

        FloatArray highs = list.getHigh();
        FloatArray lows = list.getLow();

        for(int i = span; i < size; i++)
        {
            //Conversion Line
            float high = highs.max(i-conversion+1, i);
            float low  = lows.min(i-conversion+1, i);
            float conversionLine = (high + low) / 2;

            //Base line
            high = highs.max(i-base+1, i);
            low  = lows.min(i-base+1, i);
            float baseLine = (high + low) / 2;

            //Leading Span A
            spanA.mVal[i] = (conversionLine + baseLine) / 2;

            //Leading Span B
            high = highs.max(i-span+1, i);
            low  = lows.min(i-span+1, i);
            spanB.mVal[i] = (high + low) / 2;
        }

        return new PairArray(spanA,spanB);
    }
}

File: src/main/java/org/cerion/stocklist/overlays/IchimokuClouds.kt
package org.cerion.stocklist.overlays

import org.cerion.stocklist.PriceList
import org.cerion.stocklist.arrays.FloatArray
import org.cerion.stocklist.arrays.PairArray
import org.cerion.stocklist.functions.types.PriceOverlay

class IchimokuClouds(p1: Int, p2: Int, p3: Int) : PriceOverlayBase(PriceOverlay.CLOUD, p1, p2, p3) {

    constructor() : this(9, 26, 52)

    override fun eval(list: PriceList): PairArray {
        return ichimokuCloud(list, getInt(0), getInt(1), getInt(2))
    }

    override fun getName(): String = "Ichimoku Clouds"

    private fun ichimokuCloud(list: PriceList, conversion: Int, base: Int, span: Int): PairArray {
        val size = list.size
        val spanA = FloatArray(size)
        val spanB = FloatArray(size)

        val highs = list.high
        val lows = list.low

        for (i in span until size) {
            //Conversion Line
            var high = highs.max(i - conversion + 1, i)
            var low = lows.min(i - conversion + 1, i)
            val conversionLine = (high + low) / 2

            //Base line
            high = highs.max(i - base + 1, i)
            low = lows.min(i - base + 1, i)
            val baseLine = (high + low) / 2

            //Leading Span A
            spanA.mVal[i] = (conversionLine + baseLine) / 2

            //Leading Span B
            high = highs.max(i - span + 1, i)
            low = lows.min(i - span + 1, i)
            spanB.mVal[i] = (high + low) / 2
        }

        return PairArray(spanA, spanB)
    }
}
